src/librustc_middle/ty/instance.rs

```diff
@@ -486,6 +464,42 @@ impl<'tcx> Instance<'tcx> {
             | InstanceDef::VtableShim(..) => Some(self.substs),
         }
     }
+
+    /// Returns a new `Instance` where generic parameters in `instance.substs` are replaced by
+    /// identify parameters if they are determined to be unused in `instance.def`.
+    pub fn polymorphize(self, tcx: TyCtxt<'tcx>) -> Self {
+        debug!("polymorphize: running polymorphization analysis");
+        if !tcx.sess.opts.debugging_opts.polymorphize {
+            return self;
+        }
+
+        if let InstanceDef::Item(def) = self.def {
+            let unused = tcx.unused_generic_params(def.did);
+
+            if unused.is_empty() {
+                // Exit early if every parameter was used.
+                return self;
+            }
+
+            debug!("polymorphize: unused={:?}", unused);
+            let polymorphized_substs =
+                InternalSubsts::for_item(tcx, def.did, |param, _| match param.kind {
+                // If parameter is a const or type parameter..
+                ty::GenericParamDefKind::Const | ty::GenericParamDefKind::Type { .. } if
+                    // ..and is within range and unused..
+                    unused.contains(param.index) =>
```

Comment:
Isn't this incorrect for `Item`s with more than 64 generics or am I missing something here?

`FiniteBitSet` contains returns `true` in this case:
```rust
    /// Returns `true` if the `index`th bit is set, or if `index` is outwith the range of the
    /// integral type.
    pub fn contains(&self, index: u32) -> bool {
        index >= T::DOMAIN_SIZE
            || ((self.0.checked_shr(index).unwrap_or(T::ONE)) & T::ONE) == T::ONE
    }
```

Reply:
Well spotted - fixed!